examples/vision/sam.py

```diff
@@ -0,0 +1,601 @@
+
+
+def show_masks_on_image(raw_image, masks, scores):
+    if len(masks[0].shape) == 4:
+        final_masks = tf.squeeze(masks[0])
+    if scores.shape[0] == 1:
+        final_scores = tf.squeeze(scores)
+
+    nb_predictions = scores.shape[-1]
+    fig, axes = plt.subplots(1, nb_predictions, figsize=(15, 15))
+
+    for i, (mask, score) in enumerate(zip(final_masks, final_scores)):
+        mask = tf.stop_gradient(mask)
+        axes[i].imshow(np.array(raw_image))
+        show_mask(mask, axes[i])
+        axes[i].title.set_text(f"Mask {i+1}, Score: {score.numpy().item():.3f}")
+        axes[i].axis("off")
+    plt.show()
+
+
+"""
+We will segment a car image using a point prompt. Make sure to set `return_tensors` to
+`tf` when calling the processor.
+
+Let's load an image of a car and segment it.
+"""
+
+from PIL import Image
+import requests
+
+img_url = "https://huggingface.co/ybelkada/segment-anything/resolve/main/assets/car.png"
+raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB")
+plt.imshow(raw_image)
+plt.show()
+
+"""
+Let's now define a set of points we will use as the prompt.
+"""
+
+input_points = [[[450, 600]]]
+
+# Visualize a single point.
+show_points_on_image(raw_image, input_points[0])
+
+"""
+And segment:
+"""
+
+# Preprocess the input image.
+inputs = processor(raw_image, input_points=input_points, return_tensors="tf")
+
+# Predict for segmentation with the prompt.
+outputs = model(**inputs)
+
+"""
+`outputs` has got two attributes of our interest:
+
+* `outputs.pred_masks`: which denote the predicted masks.
+* `outputs.iou_scores`: which denote the IoU scores associated with the masks.
+"""
+
+"""
+Let's post-process the masks and visulize them with their IoU scores:
+"""
+
+masks = processor.image_processor.post_process_masks(
+    outputs.pred_masks,
+    inputs["original_sizes"],
+    inputs["reshaped_input_sizes"],
+    return_tensors="tf",
+)
+
+show_masks_on_image(raw_image, masks, outputs.iou_scores)
+
+"""
+And there we go!
+
+As can be noticed, all the masks are _valid_ masks for the point prompt we provided.
+
+SAM is flexbile enough to support different visual prompts and we encourage you to check
+out [this
+notebook](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipy
+nb) to know more about them!
+"""
+
+"""
+## Fine-tuning
+
+We'll use [this dataset](https://huggingface.co/datasets/nielsr/breast-cancer) consisting
+of breast cancer scans. In the medical imaging domain, being able to segment the cells
+containing malignancy is an important task.
+
+### Data preparation
+
+Let's first get the dataset.
+"""
+
+from tensorflow import keras
+
+remote_path = "https://hf.co/datasets/sayakpaul/sample-datasets/resolve/main/breast-cancer-dataset.tar.gz"
+dataset_path = keras.utils.get_file(
+    "breast-cancer-dataset.tar.gz", remote_path, untar=True
+)
+
+"""
+Let's now visualize a sample from the dataset.
+
+*(The `show_mask()` utility is taken from [this
```

Comment:
The links should stay on a single line. So, in this case, we should do:

```
*(The `show_mask()` utility is taken from
[this notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb)
)*
```

Notice how the link is on the same line. You can follow [this source Python file](https://github.com/keras-team/keras-io/blob/master/examples/nlp/mlm_training_tpus.py) to understand this idea.

Reply:
I remember removing some of them, it's nb2py script that does it :/

Reply:
Yes, so, we need to manually remove them.